File: packages/rox_proxy_native/macos/rox_proxy_native/Sources/rox_proxy_native/Models/DomainRule.swift
import Foundation

struct DomainRule: Identifiable, Codable, Sendable, Hashable {
    let id: UUID
    var domain: String      // e.g. "api.example.com" or "*.example.com"
    var isEnabled: Bool

    init(id: UUID = UUID(), domain: String, isEnabled: Bool = true) {
        self.id = id
        self.domain = domain
        self.isEnabled = isEnabled
    }

    /// Returns true if the given host matches this rule (supports * prefix wildcard).
    func matches(host: String) -> Bool {
        guard isEnabled else { return false }
        if domain.hasPrefix("*.") {
            let suffix = String(domain.dropFirst(2))
            return host == suffix || host.hasSuffix("." + suffix)
        }
        return host == domain
    }
}

File: lib/models/domain_rule.dart
import 'package:uuid/uuid.dart';

class DomainRule {
  final String id;
  String domain;
  bool isEnabled;

  DomainRule({
    String? id,
    required this.domain,
    this.isEnabled = true,
  }) : id = id ?? const Uuid().v4();

  bool matches(String host) {
    if (!isEnabled) return false;
    if (domain.startsWith('*.')) {
      final suffix = domain.substring(2);
      return host == suffix || host.endsWith('.$suffix');
    }
    return host == domain;
  }

  Map<String, dynamic> toMap() => {
        'id': id,
        'domain': domain,
        'isEnabled': isEnabled,
      };

  factory DomainRule.fromMap(Map<String, dynamic> map) => DomainRule(
        id: map['id'] as String?,
        domain: map['domain'] as String,
        isEnabled: map['isEnabled'] as bool? ?? true,
      );
}
